wxWidgets: Convert to split package

stahta01 opened the pull request.

MehdiChinoune commented:
In general we fellow Arch Linux when splitting packages.

I couldn't find a list of rules to split packages, but I could conclude some of them:

We split :
- `-doc` when it has a huge size
- `-static` when It has a huge size.
- `-python` part when It is optional
- variants like `-qt5`/`-qt6` or `-gtk2`/`-gtk3`/`-gtk4`
- sometimes extra optional plugins.

but we never split `-dev` part (which is a DEB/RPM packaging style).

stahta01 commented:
What suffix to use when a part is optional and is only needed for development of variants?
devel instead of dev?

MehdiChinoune commented:
> What suffix to use when a part is optional and is only needed for development of variants? devel instead of dev?

We don't separate header files (development) from the core package (runtime).

stahta01 commented:
> > What suffix to use when a part is optional and is only needed for development of variants? devel instead of dev?
> 
> We don't separate header files (development) from the core package (runtime).

Then you are saying what I want to do with wxWidgets 3.2 is not allowed!!!

Tim S.

MehdiChinoune commented:
> > > What suffix to use when a part is optional and is only needed for development of variants? devel instead of dev?
> > 
> > 
> > We don't separate header files (development) from the core package (runtime).
> 
> Then you are saying what I want to do with wxWidgets 3.2 is not allowed!!!
> 
> Edit: Please identify the we you are talking about!
> 
> Tim S.

I have never seen a package on Arch or MSYS2 that was separated into runtime+development
we = MSYS2 packagers

ZachBacon commented:
> > > > What suffix to use when a part is optional and is only needed for development of variants? devel instead of dev?
> > > 
> > > 
> > > We don't separate header files (development) from the core package (runtime).
> > 
> > Then you are saying what I want to do with wxWidgets 3.2 is not allowed!!!
> > 
> > Edit: Please identify the we you are talking about!
> > 
> > Tim S.
> 
> I have never seen a package on Arch or MSYS2 that was separated into runtime+development
> we = MSYS2 packagers

I mean... Wxwidgets on arch, especially 3.2 has been split. But for development vs runtime, there's a few msys2 packages done like that, iconv is one I can think of, just not the mingw compiled iconv.

MehdiChinoune commented:
> I mean... Wxwidgets on arch, especially 3.2 has been split.

https://archlinux.org/packages/extra/x86_64/wxwidgets/
wxwidgets-common: Common libraries and headers for wxwidgets
wxwidgets-gtk3: GTK+3 implementation of wxWidgets API for GUI
wxwidgets-qt5: Qt5 implementation of wxWidgets API for GUI

libraries and headers are in the same package wxwidgets-common, the other two represent variants `-qt5` and `-gtk3`

ZachBacon commented:
> > I mean... Wxwidgets on arch, especially 3.2 has been split.
> 
> https://archlinux.org/packages/extra/x86_64/wxwidgets/
> wxwidgets-common: Common libraries and headers for wxwidgets
> wxwidgets-gtk3: GTK+3 implementation of wxWidgets API for GUI
> wxwidgets-qt5: Qt5 implementation of wxWidgets API for GUI
> 
> libraries and headers are in the same package wxwidgets-common, the other two represent variants `-qt5` and `-gtk3`

In case you missed my meaning, I did not suggest the development files were split.

stahta01 commented:
Would this naming of packages be allowed

libwxwidgets3.2-msw         Shared Libraries for wxMSW
libwxwidgets3.2-gtk3        Shared Libraries for wxGTK based on gtk3
wxwidgets3.2-gtk3           Development files for wxwidgets 3.2 based on gtk3 [wxGTK]
wxwidgets3.2-msw            Development files for wxwidgets 3.2 based on windows API [wxMSW]
wxwidgets3.2-common         Common Development files for wxWidgets 3.2
wxwidgets3.2-msw-config     wx-config for wxMSW
wxwidgets3.2-gtk3-config    wx-config for wxGTK based on gtk3

wxwidgets3.1                    wxwidgets 3.1 based on windows API [wxMSW]
wxwidgets3.1-config         wx-config for wxMSW

wxwidgets3.0                wxWidgets 3.0 based on windows API [wxMSW]
wxwidgets3.0-config         wx-config for wxWidgets 3.0
wxWidgets                   meta for wxwidgets3.0 and wxwidgets3.0-config

In time, the wxWidgets meta package will likely be changed to installing
  wxwidgets3.2-msw
  wxwidgets3.2-msw-config

Both wxwidgets3.2-gtk3 and wxwidgets3.2-msw depends on wxwidgets3.2-common

These packages all conflict with each other
  wxwidgets3.2-msw-config
  wxwidgets3.2-gtk3-config
  wxwidgets3.1-config
  wxwidgets3.0-config

I wish to have wxGTK and wxMSW based packages installed and working at the same time.

If that is against your rules I think the rules need to be changed!

I can not provide the wx-config scripts or them installed by wxwidgets3.2-msw, wxwidgets3.1, wxwidgets3.0;
but, then I would need to have the shared DLLs not in  wxwidgets3.1 and wxwidgets3.0.

Tim S.

stahta01 commented:
I give up there seems to be no way to do what I want and what your require!

Tim S.

ZachBacon commented:
Don't give up so easily, let's see what the others say. So far we haven't heard from lazka or Biswa yet.

stahta01 commented:
package_wxwidgets3.2-base-libs() {
  pkgdesc='Base libraries for wxWidgets'

package_wxwidgets3.2-msw-libs() {
  pkgdesc='Windows libraries for wxWidgets'
  depends=(wxwidgets3.2-base-libs)

package_wxwidgets3.2-gtk3-libs() {
  pkgdesc='GTK+3 libraries for wxWidgets'
  depends=(wxwidgets3.2-base-libs)

package_wxwidgets3.2-common() {
  pkgdesc='Common utils and headers for wxwidgets'
  # will have wx-config and wxrc utils

package_wxwidgets3.2-gtk3() {
  pkgdesc='GTK+3 implementation of wxWidgets API for GUI'
  depends=(wxwidgets3.2-common wxwidgets3.2-gtk3-libs)

MehdiChinoune commented:
https://wiki.archlinux.org/title/Arch_Linux#Simplicity

ZachBacon commented:
@Biswa96 @lazka can we have your thoughts please

lazka commented:
I'm having a hard time figuring out how everything fits together, but give the last example I'd merge `common` with `base-libs` and `gtk3-libs` with `gtk3`.

But, given the complexity of all those versions and backend combinations I'm fine with @stahta01 just doing what he thinks is easiest for him. Once things settle down we can always simplify things if needed.

stahta01 commented:
The problem is unless the libs are separate you can not run two wxWidgets versions or toolkits.

Example toolkits are msw, qt, gtk2, or gtk3. And, I would like to run 3.0, 3.1, and, 3.2 apps at the same time.
I could merge common with the toolkit non lib package; but, it will result in larger download.
Since, there will be two copies of everything in common.
I will start work on the package renames.

Tim S.

mmuetzel commented on mingw-w64-wxWidgets/PKGBUILD:
```diff
@@ -6,23 +6,16 @@
 mingw_arch=('mingw32' 'mingw64' 'ucrt64' 'clang64' 'clang32' 'clangarm64')
 license=("spdx:GPL-2.0-or-later AND WxWindows-exception-3.1")
 url="https://www.wxwidgets.org/"
-depends=("${MINGW_PACKAGE_PREFIX}-gcc-libs"
-         "${MINGW_PACKAGE_PREFIX}-expat"
-         "${MINGW_PACKAGE_PREFIX}-libjpeg-turbo"
-         "${MINGW_PACKAGE_PREFIX}-libpng"
-         "${MINGW_PACKAGE_PREFIX}-libtiff"
-         "${MINGW_PACKAGE_PREFIX}-xz"
-         "${MINGW_PACKAGE_PREFIX}-zlib")
```
IIUC, if you move `depends` from the base package section to a split package, you'll need to add them to `makedepends` of the base package so that they are still being installed when the runners build this rule.

stahta01 commented:
I am next going to do wxWidgets 3.1 or 3.2 packages

stahta01 commented:
Looks like I need to adjust the package conflicts between 3.0 and 3.2 packages

stahta01 commented:
Added the provides and conflicts that wxWidgets 3.2 work shows is needed.
But, wxwidgets 3.2 work is going slow; since, I removed two packages from the logical design by popular demand, I am running into installation conflicts. 

Tim S.
